dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1002 @@
+    """Acquires log messages from GCP logs for a specific project id and time
+    range.
+
+    Args:
+      project_id: Project id from which we are going to obtain the logs.
+      start_timestamp: Retrieve logs starting at this timestamp.
+      end_timestamp: Retrieve logs ending at this timestamp.
+      resource_id: Optional resource id
+
+    Return:
+      List of log messages
+    """
+    if not project_id or not start_timestamp or not end_timestamp:
+      return []
+
+    gcl = GoogleCloudLog(project_ids=[project_id])
+
+    if not self.period_covered_by_retrieved_logs.get(
+        'start') or not self.period_covered_by_retrieved_logs.get('end'):
+      self.period_covered_by_retrieved_logs['start'] = start_timestamp
+      self.period_covered_by_retrieved_logs['end'] = end_timestamp
+    else:
+      # If the required time range is within the time range already retrieved
+      # before return an empty list
+      if start_timestamp >= self.period_covered_by_retrieved_logs[
+          'start'] and end_timestamp <= self.period_covered_by_retrieved_logs[
+              'end']:
+        return []
+
+    # Make sure we only request the period that was not retrieved before, for
+    # optimization
+    if start_timestamp < self.period_covered_by_retrieved_logs['start']:
+      if end_timestamp < self.period_covered_by_retrieved_logs['end']:
+        end_timestamp = self.period_covered_by_retrieved_logs['start']
+      self.period_covered_by_retrieved_logs['start'] = start_timestamp
+
+    if end_timestamp > self.period_covered_by_retrieved_logs['end']:
+      if start_timestamp > self.period_covered_by_retrieved_logs['start']:
+        start_timestamp = self.period_covered_by_retrieved_logs['end']
+      self.period_covered_by_retrieved_logs['end'] = end_timestamp
+
+    # pylint: disable=line-too-long
+    self.PublishMessage(
+        f'Retrieving logs from { start_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")} to {end_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")}.'
+    )
+
+    query_filter = f'resource.type = ("gce_instance" OR "api" OR "gce_disk" OR "gce_image" OR "gce_instance_template" OR "gce_snapshot") \
+    AND logName = "projects/{project_id}/logs/cloudaudit.googleapis.com%2Factivity" \
+    AND operation.first = "true" \
+    AND timestamp >= "{start_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")}" \
+    AND timestamp <= "{end_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")}" \
+    AND severity=NOTICE \
+    AND protoPayload.methodName : ("insert" OR "create" OR "delete")'
+
+    if resource_id:
+      query_filter = query_filter + \
+          f' AND (resource.labels.instance_id="{resource_id}" OR resource.labels.image_id="{resource_id}")'
+
+    log_messages: List[Dict[str,
+                            Any]] = gcl.ExecuteQuery(qfilter=[query_filter])
+
+    return log_messages
+
+  def _ParseLogMessagesFromFileContainer(
```

Comment:
just out of curiosity, is the the expected file format one log entry per line? is it the default format for downloaded gcp logs?

Reply:
Om I see now, its a json file